imap-proto/src/types.rs

```diff
@@ -118,13 +118,97 @@ pub enum SectionPath {
     Part(Vec<u32>, Option<MessageSection>),
 }
 
+pub type BodyParam<'a> = Option<Vec<(&'a str, &'a str)>>;
+
+#[derive(Debug, Eq, PartialEq)]
+pub struct BodyDisposition<'a> {
+    pub disposition_type: &'a str,
+    pub param: Option<Vec<(&'a str, &'a str)>>,
+}
+
+#[derive(Debug, Eq, PartialEq)]
+pub enum BodyExtension<'a> {
+    Num(u32),
+    Str(Option<&'a str>),
+    List(Vec<BodyExtension<'a>>),
+}
+
+#[derive(Debug, Eq, PartialEq)]
+pub struct BodyStructureText<'a> {
```

Comment:
As I mention in the previous PR, there's a lot of duplication here. Should probably factor out those fields that are on every variant, and probably also the fields that are on text/basic/message but not multipart. I'm not sure what good type names are to convey the concepts here, can you make a proposal?

|   | text | basic | message | multipart |
|---|---|---|---|---|
| param | x | x | x | x |
| disposition | x | x | x | x |
| lang | x | x | x | x |
| location | x | x | x | x |
| extension | x | x | x | x |
| id | x | x | x |   |
| description | x | x | x |   |
| encoding | x | x | x |   |
| octets | x | x | x |   |
| md5 | x | x | x |   |
| envelope | | | x | |
| lines | x | x |   |   |
| bodies | | | | x |
| media_type | | x |   |   |
| media_subtype | x | x |   | x |

Reply:
How about the private types in [`body_structure.rs`](https://github.com/djc/tokio-imap/pull/48/commits/a5140492dc337b34cfc95f9c1af55609cd0ac21f#diff-2af6f4b98e4e6881c1c7dbdac302e407R9)?

Reply:
Externalizing BodyFields could make sense, but the duplication between 1Part and MPart doesn't make sense to me.